### DISCUSSION TITLE
v2: How do I correctly annotate a nullable foreign key with `Annotated` mappings?

### DISCUSSION THREAD

**Discussion by lukasjuhrich** [sqlalchemy/sqlalchemy#9176] [Usage Questions] [answered]
When using annotations like these, the `| None` optionality does not seem to make its way to the `room_fk` column:

```python
from __future__ import annotations
import typing as t

from sqalachemy import *
from sqalachemy.orm import *
from sqalachemy.schema import *


class Base(DeclarativeBase): ...

intpk = t.Annotated[int, mapped_column(primary_key=True)]
room_fk = t.Annotated[int | None, mapped_column(ForeignKey("room.id"))]


class User(Base):
    __tablename__ = "user_account"

    id: Mapped[intpk]
    name: Mapped[str]
    fullname: Mapped[str | None]
    room_id: Mapped[room_fk]
    room: Mapped[Room | None] = relationship(back_populates="inhabitants")


class Room(Base):
    __tablename__ = "room"

    id: Mapped[intpk]
    name: Mapped[str]
    inhabitants: Mapped[list[User]] = relationship(back_populates="room")

print(CreateTable(User.__table__))
```
Output:
```sql
CREATE TABLE user_account (
        id INTEGER NOT NULL, 
        name VARCHAR NOT NULL, 
        fullname VARCHAR, 
        room_id INTEGER NOT NULL, 
        PRIMARY KEY (id), 
        FOREIGN KEY(room_id) REFERENCES room (id)
)
```

---- 

_(found out while writing this answer):_
It seems like setting `room_id: Mapped[room_fk | None]` works:
```sql
CREATE TABLE user_account (
        id INTEGER NOT NULL, 
        name VARCHAR NOT NULL, 
        fullname VARCHAR, 
        room_id INTEGER, 
        PRIMARY KEY (id), 
        FOREIGN KEY(room_id) REFERENCES room (id)
)
```

Is this by design, or an accident? Ignoring the nullability in the `Annotated` field seems like something that could easily go unnoticed.

I think arrived at this because I assumed the following rewritings would have equivalent semantics:
```python
# rewriting 1: from…
col: Mapped[pytype] = mapped_column(**kw)
# …to
col: Mapped[Annotated[pytype, mapped_column(**kw)]]

# rewriting 2: from…
class Foo (Base):
    ...
    col: Mapped[py_expr]
# …to
alias = py_expr
class Foo(Base):
    ...
    col: Mapped[alias]
```

**Comment by zzzeek** [comment, marked as answer]
the nullability-from-annotations logic is extremely crazy since there are so many left turns in it, so this particular case likely wasn't considered, and/or there's some element of it that doesn't actually work in practice (in which case it would be better if it emitted a warning)

for now it's most striaghtforward for your case to just add the nullability explicitly:


    room_fk = t.Annotated[int | None, mapped_column(ForeignKey("room.id"), nullable=True)]

I'd be surprised if *that* didn't work

**Reply by zzzeek** [reply]
anyway this is #9177
